lib/functions/configuration/main-config.sh

```diff
@@ -168,80 +168,77 @@ function do_main_configuration() {
+			declare -g -r MAINLINE_FIRMWARE_SOURCE='https://mirrors.tuna.tsinghua.edu.cn/git/linux-firmware.git'
 			;;
 		bfsu)
-			MAINLINE_KERNEL_SOURCE='https://mirrors.bfsu.edu.cn/git/linux-stable.git'
-			MAINLINE_FIRMWARE_SOURCE='https://mirrors.bfsu.edu.cn/git/linux-firmware.git'
+			declare -g -r MAINLINE_KERNEL_SOURCE='https://mirrors.bfsu.edu.cn/git/linux-stable.git'
+			declare -g -r MAINLINE_FIRMWARE_SOURCE='https://mirrors.bfsu.edu.cn/git/linux-firmware.git'
 			;;
 		*)
-			MAINLINE_KERNEL_SOURCE='https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git' # "linux-stable" was renamed to "linux"
-			MAINLINE_FIRMWARE_SOURCE='https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git'
+			declare -g -r MAINLINE_KERNEL_SOURCE='https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git' # "linux-stable" was renamed to "linux"
+			declare -g -r MAINLINE_FIRMWARE_SOURCE='https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git'
 			;;
 	esac
 
-	MAINLINE_KERNEL_DIR='linux-mainline'
 
-	[[ $USE_GITHUB_UBOOT_MIRROR == yes ]] && UBOOT_MIRROR=github
+	[[ $USE_GITHUB_UBOOT_MIRROR == yes ]] && UBOOT_MIRROR=github # legacy compatibility?
 
 	case $UBOOT_MIRROR in
 		gitee)
-			MAINLINE_UBOOT_SOURCE='https://gitee.com/mirrors/u-boot.git'
+			declare -g -r MAINLINE_UBOOT_SOURCE='https://gitee.com/mirrors/u-boot.git'
 			;;
 		denx)
-			MAINLINE_UBOOT_SOURCE='https://source.denx.de/u-boot/u-boot.git'
+			declare -g -r MAINLINE_UBOOT_SOURCE='https://source.denx.de/u-boot/u-boot.git'
 			;;
 		*)
-			MAINLINE_UBOOT_SOURCE='https://github.com/u-boot/u-boot'
+			declare -g -r MAINLINE_UBOOT_SOURCE='https://github.com/u-boot/u-boot'
 			;;
 	esac
 
-	MAINLINE_UBOOT_DIR='u-boot'
-
 	case $GITHUB_MIRROR in
 		fastgit)
-			GITHUB_SOURCE='https://hub.fastgit.xyz'
+			declare -g -r GITHUB_SOURCE='https://hub.fastgit.xyz'
 			;;
 		ghproxy)
 			[[ -z $GHPROXY_ADDRESS ]] && GHPROXY_ADDRESS=mirror.ghproxy.com
-			GITHUB_SOURCE="https://${GHPROXY_ADDRESS}/https://github.com"
+			declare -g -r GITHUB_SOURCE="https://${GHPROXY_ADDRESS}/https://github.com"
 			;;
 		gitclone)
-			GITHUB_SOURCE='https://gitclone.com/github.com'
+			declare -g -r GITHUB_SOURCE='https://gitclone.com/github.com'
 			;;
 		*)
-			GITHUB_SOURCE='https://github.com'
+			declare -g -r GITHUB_SOURCE='https://github.com'
 			;;
 	esac
 
 	case $GHCR_MIRROR in
 		dockerproxy)
-			GHCR_SOURCE='ghcr.dockerproxy.com'
+			declare -g -r GHCR_SOURCE='ghcr.dockerproxy.com'
 			;;
 		*)
-			GHCR_SOURCE='ghcr.io'
+			declare -g -r GHCR_SOURCE='ghcr.io'
 			;;
 	esac
 
 	# Let's set default data if not defined in board configuration above
 	[[ -z $OFFSET ]] && OFFSET=4 # offset to 1st partition (we use 4MiB boundaries by default)
-	[[ -z $ARCH ]] && ARCH=armhf # makes little sense to default to anything...
-	ATF_COMPILE=yes
+	[[ -z $ARCH ]] && ARCH=armhf # makes little sense to default to anything... # @TODO: remove
```

Comment:
Probably best to just error out if not set.

Reply:
Yeah this is a long-standing problem we'd better handle separately -- I just added todo.